htlcswitch/link.go

```diff
@@ -797,12 +794,61 @@ func (l *channelLink) htlcManager() {
 	if l.cfg.SyncStates {
 		err := l.syncChanStates()
 		if err != nil {
-			l.errorf("unable to synchronize channel states: %v", err)
-			if err != ErrLinkShuttingDown {
-				// TODO(halseth): must be revisted when
-				// data-loss protection is in.
-				l.fail(LinkFailureError{code: ErrSyncError},
-					err.Error())
+			switch {
+			case err == ErrLinkShuttingDown:
+				log.Debugf("unable to sunc channel states, " +
+					"link is shutting down")
+				break
+
+			case err == lnwallet.ErrCommitSyncRemoteDataLoss:
+				// We failed syncing the commit chains,
+				// probably because the remote has lost
+				// state. We should force close the channel.
+				l.fail(
+					LinkFailureError{
+						code:       ErrSyncError,
+						ForceClose: true,
+					},
+					"unable to synchronize channel "+
+						"states: %v", err,
+				)
+				break
+
+			case err == lnwallet.ErrCommitSyncLocalDataLoss:
+				// We have lost state and cannot safely force
+				// close the channel. Fail the channel and wait
+				// for the remote to hopefully force close it.
+				// The remote has sent us its latest unrevoked
+				// commitment point, that we stored in the
+				// database, that we can use to retrieve the
+				// funds when the remote closes the channel.
```

Comment:
Style nit on these comments: typically we place them right above the `case` statement. A bonus of this is that the blocks are shorter as there's less line wrapping needed.